### PR TITLE
Experiment object

### PR DESCRIPTION
**Pull request by jindrahelcl**
A lot of refactoring + changes of some concepts

Introduced changes:
- Entrypoints: Specified in the ini file, the entrypoints are objects with an `execute` function, which is called after parsing the configuration file. Are contained in `neuralmonkey/entrypoints` package.
- `Experiment` and `Model` objects. Both are entrypoints:
  - model stores the architecture of the network, has methods `feed_dicts`, `run_on_dataset` and `execute`. 
  - experiment stores the training parameters and its `execute` method performs the training.
  - both of these are written in a way the actual model parameters are a not part of these objects.
- Single binary: The new way to run Neuralmonkey is to call `nmonkey COMMAND INI_FILE [ARGS...]`   `nmonkey` is because `neuralmonkey` is a link to the package (we can solve this later). The `COMMAND` corresponds to a section in the ini file, which must represent an `Entrypoint` object.
- No `learning_utils`: The training loop is now inside the module `entrypoints.experiment`.
- `OutputDirectory`: Helper class that deals with file names
- Allow config objects to be from elsewhere than neuralmonkey package
  - if module exists in neuralmonkey package, it is used, otherwise, the parser tries to import it without adding the `neuralmonkey.` prefix
  - TensorFlow's ConfigProto can be constructed this way (see `tests/small.ini`)

Issues related to this change:
#9, #17, #38, #61, #66, #67, #71 and #72

To be done:
- Server binary + enabling tests
- Names for evaluation functions
- Better handling of the `tfconfig` parameter (now it's needed both in run and  train entrypoints)

### CODE REVIEW

**Comment by jlibovicky** [comment]
Je super, že jsi s tím takhle pohnul. Mám milion komentářů. Proberem to osobně v sobotu, v neděli atd. :-)

**Comment by jindrahelcl** [comment]
cool :)

**Comment by jindrahelcl** [comment]
Hlavně to nefunguje. Ale nevim, jestli to zkazil tenhle refactor, nebo jestli to byl ten refactor toho dekodéru..

**Comment by jindrahelcl** [comment]
Dekoder opraven v #76

**Comment by jindrahelcl** [comment]
Berlínský refactor jsem zatím neřešil, ale za to jsem napsal ensembly, který nejspíš budou fungovat. Zkus se na to mrknout, @jlibovicky :-)

**Comment by jlibovicky** [comment]
OK, podívám. V týhle větvi jsou ty ensembly? Už jsem úplně zapomněl, jak je to tady vymyšlený s tim experiment objektem.

Jak se má tahle větev k tý maratonský větvi?

Berlínský refaktor udělám klidně sám přespříští týden, až budu zpátky.

**Comment by jindrahelcl** [comment]
Jo, jsou v týhle větvi. Použil jsem jí, protože se mi v ní o nich snáz přemejšlelo.. Nakonec by se ty ansambly daly asi stejným způsobem udělat i v masteru. Hlavní idea je, že runner žere pole sessions, každá inicializovaná svým souborem a pak se sčítaj pravděpodobnosti každýho modelu v každym času (podle slidů z ACL tutoriálu [zde](http://nlp.stanford.edu/projects/nmt/Luong-Cho-Manning-NMT-ACL2016-v4.pdf), str. 96. Graf je sdílenej. Jak to bude s pamětí nevim, ale při tréninku se to používat nebude, tak to v nejhorším může sedět na cpu. Je to i přidaný do testů.

Teď jsem si vzpomněl, že sčítám logaritmy pravděpodobností, takže to bude odpovídat spíš tomu consensus-building scheme. Daly by se sčítat i samotný pravděpodobnosti. Normalizaci dělá dekodér, v každym čase by se to přes slovník mělo sečíst do 1 (resp. do 0, počítaj-li se logproby)...

**Comment by jindrahelcl** [comment]
K maratonský větvi se nemá nijak.
